File: js/src/parse.js
import Parsimmon from 'parsimmon';
import AST from './ast';
import Symbol from './symbol';

let { string, regex, optWhitespace, seq, lazy, alt, succeed, fail } = Parsimmon;

function lexeme(p) { return p.skip(optWhitespace); }

let FUN  = lexeme(string("fun"));
let LET  = lexeme(string("let"));
let REC  = lexeme(string("rec"));
let IN   = lexeme(string("in"));
let IF   = lexeme(string("if"));
let THEN = lexeme(string("then"));
let ELSE = lexeme(string("else"));
let LPAR  = lexeme(string("("));
let RPAR  = lexeme(string(")"));
let EQ    = lexeme(string("="));

function mkIdent(id) {
  return AST.Ide(new Symbol(id));
}

let failReserved = alt(FUN, LET, REC, IN, IF, THEN, ELSE).chain(() => fail("reserved word not expected"))

// Manually eliminating left recursion on rule: Expr -> Expr ( Expr )
let expr = lazy("an expression", function() {
  return alt(lambda, cond, letExpr, parenExpr, atom, number).chain(function(lhs) {
    function apply(l, r) { return AST.Apply(l, r) }
    return parenExpr.many().map((r) => r.reduce(apply, lhs));
  })
});
let decl = lazy("a declaration", () => alt(recursion, parenDecl, assign));

let number    = lexeme(regex(/[0-9]+/).map(mkIdent));
let atom      = lexeme(regex(/[a-z_]\w*/i)).map(mkIdent);
let lambda    = seq(FUN, LPAR, atom, RPAR, expr).map((r) => AST.Lambda(r[2], r[4]));
let cond      = seq(IF, expr, THEN, expr, ELSE, expr).map((r) => AST.Cond(r[1], r[3], r[5]));
let letExpr   = seq(LET, decl, IN, expr).map((r) => AST.Block(r[1], r[3]));
let parenExpr = seq(LPAR, expr, RPAR).map((r) => r[1]);

let assign    = seq(atom, EQ, expr).map((r) => AST.Def(r[0].ide, r[2]));
let recursion = seq(REC, decl).map((r) => AST.Rec(r[1]));
let parenDecl = seq(LPAR, decl, RPAR).map((r) => r[1]);

export default function parse(input) {
  let result = optWhitespace.then(expr).parse(input);
  if (result.value) {
    return result.value;
  }

  console.log(Parsimmon.formatError(input, result))
  throw new Error();
}

File: hs/src/Parse.hs
module Parse(parseExpr, parseFun) where

import Control.Applicative hiding (many, optional, (<|>))
import Control.Monad
import Text.Parsec
import Text.Parsec.String
import Text.Parsec.Expr
import qualified Text.Parsec.Token as T
import Text.Parsec.Language (emptyDef)

import AST

funStyle :: T.LanguageDef st
funStyle = emptyDef{ T.commentStart = "{-"
                   , T.commentEnd = "-}"
                   , T.identStart = letter
                   , T.identLetter = alphaNum
                   , T.reservedOpNames = ["="]
                   , T.reservedNames = ["fun", "let", "rec", "in",
                                        "if", "then", "else"]
                   }

-- Exp ::=
--   Ide |
--   "if" Exp "then" Exp "else" Exp |
--   "fun" "(" Ide ")" Exp |
--   Exp "(" Exp ")" |
--   "let" Decl "in" Exp | "(" Exp ")"
-- Decl ::=
--   Ide "=" Exp |
--   Decl "then" Decl | "rec" Decl |
--   "(" Decl ")"

-- http://stackoverflow.com/a/26539615/183140
lexer = T.makeTokenParser funStyle
parens = T.parens lexer
identifier = T.identifier lexer
integer = T.integer lexer
reserved = T.reserved lexer
reservedOp = T.reservedOp lexer
symbol = T.symbol lexer

atom = liftM Var identifier
number = liftM Num integer

cond :: Parser Expr
cond = do
  reserved "if"
  test <- expr
  reserved "then"
  ifTrue <- expr
  reserved "else"
  ifFalse <- expr
  return $ Cond test ifTrue ifFalse

lambda :: Parser Expr
lambda = do
  reserved "fun"
  var <- parens identifier
  body <- expr
  return $ Lambda var body

block :: Parser Expr
block = do
  reserved "let"
  d <- decl
  reserved "in"
  e <- expr
  return $ Block d e

term :: Parser Expr
term = let simpleTerm = parens(expr)
                    <|> cond
                    <|> block
                    <|> lambda
                    <|> number
                    <|> atom
       in do
         t <- simpleTerm
         applications <- many (parens exprList)
         return $ foldl FunCall t $ concat applications

exprList :: Parser [Expr]
exprList = expr `sepBy` (symbol ",")

mkUnOp op a = FunCall (Var op) a
mkBinOp op a b = FunCall (mkUnOp op a) b

binary name fn = Infix (do { reservedOp name; return (mkBinOp fn) }) AssocLeft
unary name fn = Prefix (do { reservedOp name; return (mkUnOp fn) })

table = [
  [ unary "-" "negate" ],
  [ unary "!" "not" ],
  [ binary "&" "and", binary "|" "or" ],
  [ binary "*" "times", binary "/" "div" ],
  [ binary "+" "plus", binary "-" "minus" ],
  [ binary "==" "eq" ]]

expr :: Parser Expr
expr = buildExpressionParser table term <?> "expression"

assign :: Parser Decl
assign = do
  var <- identifier
  reservedOp "="
  e <- expr
  return $ Assign var e

seqDecl :: Parser Decl
seqDecl = liftM2 Seq decl decl

rec :: Parser Decl
rec = liftM Rec (reserved "rec" *> decl)

decl :: Parser Decl
decl = let simpleDecl = assign <|> rec <|> parens(decl)
       in do
         t <- simpleDecl `sepBy` (reserved "then")
         return $ foldl1 Seq t

parseExpr :: String -> Either ParseError Expr
parseExpr str = parse (expr <* eof) "" str

parseFun :: String -> Expr
parseFun str = case parseExpr str of
                Left err -> error $ show err
                Right expr -> expr
